sparse/file.go

```diff
@@ -33,195 +31,56 @@ func SetupFileIO(direct bool) {
 		fileOpen = fio.OpenFile
 		fileReadAt = fio.ReadAt
 		fileWriteAt = fio.WriteAt
-		log.Info("Mode: directfio")
+		log.Debug("Mode: directfio")
 	} else {
 		fileOpen = fileBufferedOpen
 		fileReadAt = fileBufferedReadAt
 		fileWriteAt = fileBufferedWriteAt
-		log.Info("Mode: buffered")
+		log.Debug("Mode: buffered")
 	}
 }
 
-func loadFileLayout(abortStream <-chan error, file *os.File, layoutStream chan<- FileInterval, errStream chan<- error) error {
-	size, err := file.Seek(0, os.SEEK_END)
+func ReadDataInterval(file *os.File, dataInterval Interval) ([]byte, error) {
+	log.Debug("reading data from file ...")
+	data := make([]byte, dataInterval.Len())
+	n, err := fileReadAt(file, data, dataInterval.Begin)
 	if err != nil {
-		return err
-	}
-	go RetrieveLayoutStream(abortStream, file, Interval{0, size}, layoutStream, errStream)
-	return nil
-}
-
-// IntervalSplitter limits file intervals to predefined batch size
-func IntervalSplitter(spltterStream <-chan FileInterval, fileStream chan<- FileInterval) {
-	const batch = 32 * Blocks
-	for r := range spltterStream {
-		if verboseServer {
-			log.Debug("Interval Splitter:", r)
-		}
-		switch r.Kind {
-		case SparseHole:
-			// Process hole
-			fileStream <- r
-		case SparseData:
-			// Process data in chunks
-			for offset := r.Begin; offset < r.End; {
-				size := batch
-				if offset+size > r.End {
-					size = r.End - offset
-				}
-				interval := Interval{offset, offset + size}
-				if size == batch && interval.End%batch != 0 {
-					interval.End = interval.End / batch * batch
-				}
-				log.Debug("Interval Splitter data:", interval)
-				fileStream <- FileInterval{SparseData, interval}
-				offset += interval.Len()
-			}
+		if err == io.EOF {
+			log.Debug("have read at the end of file, total read: ", n)
+		} else {
+			errStr := fmt.Sprintf("File read error: %s", err)
+			log.Error(errStr)
+			return nil, fmt.Errorf(errStr)
 		}
+	} else if int64(n) != dataInterval.Len() {
+		errStr := "File read underrun"
```

Comment:
Nice catch. This actually will never happen, the os.ReadAt will either return the len of the slice passed in, or return EOF, which is handled already. So I should remove this else if section.